[cds-plugin-ui5] running mbt build with @sap/cds-dk@8 gives an error: Invalid priority value for UI5Plugin

**Describe the bug**
We are using `cds-plugin-ui5` and we where running `mbt build` and where greeted with an error today (15-07-2024). It was working before, the last time we tried it was 11-07-2024 and it worked (locally and via the azure pipeline).

After a lot of trying we found it's because of this command in our mta.yaml: `- npx -p @sap/cds-dk cds build --production` when we change it to `- npx -p @sap/cds-dk@7 cds build --production`, notice the `@7` after `cds-dk` package it builds with success.

We got the same error in our azure devops pipeline, luckily we could recreate locally.

**To Reproduce**
Steps to reproduce the behavior:
1. We have an `MTA` application, it consist of cds module and a fiori module with an mta.yaml file
2. In the mta.yaml we have the following:
```yaml
...
build-parameters:
  before-all:
  - builder: custom
    commands:
    - npx -p @sap/cds-dk cds build --production
```
4. Via the terminal, we used `mbt build` command and we got an error.
5. See error

```bash
[2024-07-15 17:06:39]  INFO Cloud MTA Build Tool version 1.2.30
[2024-07-15 17:06:39]  INFO generating the "Makefile_20240715170639.mta" file...
[2024-07-15 17:06:39]  INFO done
[2024-07-15 17:06:39]  INFO executing the "make -f Makefile_20240715170639.mta p=cf mtar= strict=true mode=" command...
[2024-07-15 17:06:39]  INFO validating the MTA project
[2024-07-15 17:06:39]  INFO running the "before-all" build...
[2024-07-15 17:06:39]  INFO executing the "npx -p @sap/cds-dk cds build --production" command...
building project [/home/user/projects/SAP-BTP-GZ-PROJECTS], clean [true]
cds-dk [8.0.2], cds [7.9.3], compiler [4.9.4], home [/home/user/projects/SAP-BTP-GZ-PROJECTS/node_modules/@sap/cds]

{
  build: {
    target: 'gen',
    tasks: [
      { for: 'hana', src: 'db', options: { model: ['db', 'srv', 'app'] }},
      { for: 'nodejs', src: 'srv', options: { model: ['db', 'srv', 'app'] }},
      { for: 'ui5', src: 'srv', options: { model: ['db', 'srv', 'app'] }}
    ]
  }
}

Error: Invalid priority value for UI5Plugin.
The valid priority value range for custom plugins is -1024..-1 and 512..1024.
Range 0..512 is blocked for internal plugins.
The higher the value the earlier the plugin is run.
    at BuildTaskEngine._createPlugin (/home/user/.npm/_npx/b993340195f82cd5/node_modules/@sap/cds-dk/lib/build/buildTaskEngine.js:212:19)
    at /home/user/.npm/_npx/b993340195f82cd5/node_modules/@sap/cds-dk/lib/build/buildTaskEngine.js:49:37
    at Array.forEach (<anonymous>)
    at BuildTaskEngine.processTasks (/home/user/.npm/_npx/b993340195f82cd5/node_modules/@sap/cds-dk/lib/build/buildTaskEngine.js:47:15)
    at async build (/home/user/.npm/_npx/b993340195f82cd5/node_modules/@sap/cds-dk/lib/build/index.js:19:16)
    at async Object.build (/home/user/.npm/_npx/b993340195f82cd5/node_modules/@sap/cds-dk/bin/build.js:171:5)
    at async Object.exec (/home/user/.npm/_npx/b993340195f82cd5/node_modules/@sap/cds-dk/bin/cds.js:81:16)
[2024-07-15 17:06:40] ERROR the "before-all"" build failed: could not execute the "npx -p @sap/cds-dk cds build --production" command: exit status 1
make: *** [Makefile_20240715170639.mta:28: pre_build] Error 1
[2024-07-15 17:06:40] ERROR could not build the MTA project: could not execute the "make -f Makefile_20240715170639.mta p=cf mtar= strict=true mode=" command: exit status 2
Error: could not build the MTA project: could not execute the "make -f Makefile_20240715170639.mta p=cf mtar= strict=true mode=" command: exit status 2
```

**Expected behavior**
A successful `mbt build` with @sap/cds-dk@8.

**Desktop (please complete the following information):**
 - OS: A snippet from `inxi`
 ```bash
System:
  Kernel: 6.6.34-1-MANJARO arch: x86_64 bits: 64 compiler: gcc v: 14.1.1
  Desktop: KDE Plasma v: 6.0.5 Distro: Manjaro base: Arch Linux
```
 - Package.json: a snippet
 ```json
...
"dependencies": {
        "@azure/storage-blob": "12.23.0",
        "@cap-js-community/odata-v2-adapter": "1.12.11",
        "@esri/arcgis-rest-feature-layer": "3.7.0",
        "@sap-cloud-sdk/http-client": "3.16.0",
        "@sap-cloud-sdk/util": "3.16.0",
        "@sap/cds": "7.9.3",
        "@sap/cds-hana": "2.0.0",
        "@sap/textbundle": "5.1.0",
        "@sap/xssec": "4.1.3",
        "base64-stream": "1.0.0",
        "decimal.js": "10.4.3",
        "email-templates": "11.1.1",
        "express": "4.19.2",
        "hdb": "0.19.8",
        "luxon": "3.4.4",
        "passport": "0.7.0",
        "pdfmake": "0.2.10",
        "sap-cf-mailer": "0.0.5"
    },
    "devDependencies": {
        "@cap-js/sqlite": "1.7.2",
        "cds-plugin-ui5": "0.9.3",
        "cds-swagger-ui-express": "0.8.0",
        "eslint": "9.6.0",
        "jest": "29.7.0",
        "jest-jasmine2": "29.7.0",
        "jest-junit": "16.0.0",
        "rimraf": "5.0.7",
        "supertest": "7.0.0"
    },
...
"cds": {
        "cds-plugin-ui5": {
            "modules": {
                "fiorimodule": {
                    "configFile": "ui5.yaml",
                    "mountPath": "/fiorimodule/webapp"
                }
            }
        },
...
```

**Additional context**
Is this the right place to report this issue or does this need to be reported via SAP for me and make a case of it? If you need more information, let me know!

--- a/packages/cds-plugin-ui5/cds-plugin.js
+++ b/packages/cds-plugin-ui5/cds-plugin.js
@@ -48,6 +48,35 @@ if (!skip) {
 	// to disable the ui5-middleware-cap if used in apps
 	process.env["cds-plugin-ui5"] = true;
 
+	const { execSync } = require("child_process");
+	const { readFileSync } = require("fs");
+
+	// function to resolve a module with the given paths without throwing an error
+	const resolveModule = function resolveModule(moduleName, paths) {
+		try {
+			return require.resolve(moduleName, { paths });
+		} catch (error) {
+			return null;
+		}
+	};
+
+	// find out the CDS-DK version to control the behavior of the plugin
+	const getCDSDKVersion = function getCDSDKVersion() {
+		const moduleName = "@sap/cds-dk";
+		const globalModulesPath = execSync("npm root -g").toString().trim();
+		let resolvedPath = resolveModule(`${moduleName}/package.json`);
+		if (!resolvedPath) {
+			resolvedPath = resolveModule(`${moduleName}/package.json`, [globalModulesPath]);
+		}
+		if (resolvedPath) {
+			const packageJson = JSON.parse(readFileSync(resolvedPath, { encoding: "utf-8" }));
+			return packageJson.version;
+		}
+	};
+
+	// get the CDS-DK version to control the behavior of the plugin
+	const cdsdkVersion = getCDSDKVersion();
+
 	// promise to await the bootstrap and lock the
 	// served event to delay the startup a bit
 	let bootstrapped;
@@ -194,11 +223,11 @@ if (!skip) {
 	// check if the register function for build tasks is available at the cds object
 	// and if the Plugin class is available to register the cds build task to cover
 	// the tracks of the cds-plugin-ui5 workspace configuration and dependencies
+	const { minVersion, satisfies } = require("semver");
 	if (typeof cds.build?.register === "function" && typeof cds.build?.Plugin?.constructor === "function") {
 		const { readFile, writeFile } = require("fs").promises;
 		const { existsSync } = require("fs");
 		const { join } = require("path");
-		const { minVersion } = require("semver");
 		const util = require("util");
 		const exec = util.promisify(require("child_process").exec);
 
@@ -226,10 +255,16 @@ if (!skip) {
 				}
 				init() {}
 				clean() {
-					this._priority = -1; // hack to ensure that the task is executed last!
+					if (!satisfies(cdsdkVersion, ">=8")) {
+						this._priority = -1; // hack to ensure that the task is executed last!
+					}
 				}
 				get priority() {
-					return this._priority || 1;
+					if (!satisfies(cdsdkVersion, ">=8")) {
+						return this._priority || 1;
+					} else {
+						return -1;
+					}
 				}
 				async build() {
 					// determine the namespace from the model
@@ -277,8 +312,10 @@ if (!skip) {
 			}
 		);
 	} else {
-		// TODO: add error message to inform the user that the cds build task is not available
-		//       and that the @sap/cds-dk version is too old to support the cds build task
-		//log.info("The cds build task requires @sap/cds-dk version >= 7.6.0! Skipping execution as your @sap/cds-dk version is too old...");
+		if (!satisfies(cdsdkVersion, ">=7.5.0")) {
+			// TODO: add error message to inform the user that the cds build task is not available
+			//       and that the @sap/cds-dk version is too old to support the cds build task
+			log.warn("The cds build task requires @sap/cds-dk version >= 7.5.0! Skipping execution as your @sap/cds-dk version is too old...");
+		}
 	}
 }

